src/runtime/gc_custom.go

```diff
@@ -0,0 +1,49 @@
+//go:build gc.custom
+// +build gc.custom
+
+package runtime
+
+// This GC strategy allows an external GC to be plugged in instead of the builtin
+// implementations.
+//
+// The custom implementation must provide the following functions in the runtime package
+// using the go:linkname directive:
+//
+// - func initHeap()
+// - func alloc(size uintptr, layout unsafe.Pointer) unsafe.Pointer
+// - func free(ptr unsafe.Pointer)
+// - func GC()
+// - func KeepAlive(x interface{})
+// - func SetFinalizer(obj interface{}, finalizer interface{})
+//
+// In addition, if targeting wasi, the following functions should be exported for interoperability
+// with wasi libraries that use them. Note, this requires the export directive, not go:linkname.
+//
+// - func malloc(size uintptr) unsafe.Pointer
+// - func free(ptr unsafe.Pointer)
+// - func calloc(nmemb, size uintptr) unsafe.Pointer
+// - func realloc(oldPtr unsafe.Pointer, size uintptr) unsafe.Pointer
+
+import (
+	"unsafe"
+)
+
+func initHeap()
+
+func alloc(size uintptr, layout unsafe.Pointer) unsafe.Pointer
+
+func free(ptr unsafe.Pointer)
+
+func GC()
+
+func KeepAlive(x interface{})
+
+func SetFinalizer(obj interface{}, finalizer interface{})
+
+func setHeapEnd(newHeapEnd uintptr) {
+	// Heap is in custom GC so ignore for when called from wasm initialization.
+}
+
+func markRoots(start, end uintptr) {
+	// GC manages roots so ignore to allow gc_stack_portable to compile
```

Comment:
Some shuffling of code into separate files with build tags could remove the need for these stubs but it didn't seem worth it

Reply:
This is unsafe. The GC _must_ call `markStack` which will in turn call `markRoots`. Otherwise the stack won't be scanned for objects. The reason this call is ignored in `-gc=leaking` and `-gc=none` is because they don't deallocate anything anyway.

Reply:
The custom GC can access the stack pointer and mark the stack without calling this function itself. In the case of the bdwgc implementation, it's C code to read the stack via a pointer to a local variable seems to work fine, and if wasn't our code would crash quickly as for other cases of missing the stack.

More abstractly, when using a library like bdwgc there is no API to mark - just add roots (not needed for the stack as above though). So there isn't really a way to wire this to it.

Reply:
> In the case of the bdwgc implementation, it's C code to read the stack via a pointer to a local variable

I don't think this is correct for WebAssembly. Yes, this works for any other architecture but there is a reason why we need to keep this custom shadow stack in TinyGo for WebAssembly: WebAssembly can put values on the VM evaluation stack which isn't accessible from outside that particular function.
If bdwgc somehow manages to read the stack without this, please let me know how it is able to do this. Because I'm pretty sure this is fundamentally impossible.

>  seems to work fine, and if wasn't our code would crash quickly as for other cases of missing the stack.

No. I've found GC bugs where some roots weren't marked after months of use, so "it doesn't crash" isn't enough proof it works fine.

Reply:
The shadow stack is managed by copying to the C stack right? That's why it works, since the C stack is marked.

Reply:
We see that thanks to recent changes, that's all markStack does in TinyGo too

https://github.com/tinygo-org/tinygo/blob/release/src/runtime/gc_stack_raw.go#L19

It is just marking the C stack. That's what bdwgc does too and any custom GC would be doing it. There doesn't seem to be a need for an explicit call to do it if the C library implementing GC does it automatically

Reply:
As far as I can tell the main concern is still the volatile load - it's not so much that markStack is needed for finding the stack but that it is needed to guarantee the shadow stack is fresh. Implementations can choose to ignore the call to `addRoots` if they don't need it, as with bdwgc.

I went ahead and reverted the volatile load change and added addRoots here, which with bdwgc will be no-op. Hopefully that makes sense

Reply:
Hmm, you are right. It is indeed enough to force the compiler to have an up-to-date shadow stack by doing the volatile load. Then it can scan the globals conservatively and find the current stack that way.